src/lib/HST/Environment/LookupOrReport.hs

```diff
@@ -19,51 +21,68 @@ import qualified HST.Frontend.Syntax    as S
 import           HST.Util.Messages      ( Severity(Error), message )
 import           HST.Util.PrettyName    ( prettyName )
 
--- | Looks up the constructors belonging to the the given data type name in the
---   current environment.
+-- | Looks up the data constructor entries belonging to the the given data type
+--   name in the current environment and qualifies them so that they are
+--   unambiguous.
 --
 --   Reports a fatal error if the data type could not be found or if it could
 --   be found in multiple modules. In the latter case, the names of
---   these modules are displayed as part of the error message.
+--   these modules are displayed as part of the error message. A fatal error is
+--   also reported if not all constructors found could be identified
+--   unambiguously.
 lookupConEntriesOrReport
   :: Members '[Env a, Report] r => TypeName a -> Sem r [ConEntry a]
 lookupConEntriesOrReport typeName = do
   conEntriess <- inEnv $ lookupConEntries typeName
   case conEntriess of
-    []                -> reportFatal
+    [] -> reportFatal
       $ message Error (S.getSrcSpan typeName)
       $ "Data type not in scope: " ++ prettyName typeName
-    [(_, conEntries)] -> return conEntries
-    _                 -> reportFatal
+    [(_, Nothing)] -> reportFatal
+      $ message Error (S.getSrcSpan typeName)
+      $ "Constructor entries for the data type `"
+      ++ prettyName typeName
+      ++ "` could not be returned, because at least one of them cannot be "
+      ++ "identified unambiguously."
+    [(_, Just conEntries)] -> return conEntries
+    _ -> reportFatal
       $ message Error (S.getSrcSpan typeName)
       $ "Ambiguous data type `"
       ++ prettyName typeName
       ++ "` could refer to one of the following modules: "
       ++ displayModuleNames conEntriess
```

Comment:
If there are multiple imports for the same module, for example

```haskell
import A
import A as A'
```

a lookup of an unqualified identifier that is exported by `A` is currently considered ambiguous.

Reply:
Okay, I think this can be solved pretty easily by using
`nubBy (on (==) (interfaceName . snd)) (possibleInterfaces qName env)`
in the `lookupWith` function. I can't think of any case where two module interfaces could have the same name, so this should work.
However, we would in this case only try to qualify lookup results based on the first import declaration, which could mean that we would call some types or constructors impossible to identify unambiguously when it actually might be possible (think about adding `import B as A` to your example). However, unless you have a good solution for this, I'm not sure if we should try to solve this problem. It should occur very rarely if the developer doesn't try to cause it on purpose.

Reply:
This should now be fixed WITH correct qualification. In order to achieve that, the environment now stores a list of import declarations with each module interface.